lib/mixpanel/client.ex

```diff
@@ -35,53 +45,157 @@ defmodule Mixpanel.Client do
     GenServer.cast(process || __MODULE__, {:engage, event})
   end
 
-  def init({:ok, config}) do
-    {:ok, Enum.into(config, %{})}
+  def init({:ok, opts}) do
+    config = Enum.into(opts, @defaults)
+
+    state = %{
+      track: Queue.new(config.max_queue_track),
+      engage: Queue.new(config.max_queue_engage),
+      track_dropped: 0,
+      engage_dropped: 0
+    }
+
+    {:ok, {config, state}}
   end
 
   # No events submitted when env configuration is set to false.
-  def handle_cast(_request, %{active: false} = state) do
+  def handle_cast(_request, {%{active: false}, _} = state) do
     {:noreply, state}
   end
 
-  def handle_cast({:track, event, properties}, %{token: token} = state) do
-    data =
-      %{event: event, properties: Map.put(properties, :token, token)}
-      |> Poison.encode!()
-      |> :base64.encode()
+  def handle_cast({:track, _event, _properties} = event, {config, state}) do
+    case Queue.push(state.track, event) do
+      :dropped ->
+        new_state = Map.update!(state, :track_dropped, &(&1 + 1))
+        {:noreply, {config, new_state}, 0}
+
+      {:ok, queue} ->
+        timeout =
+          if Queue.length(queue) >= config.batch_size do
+            0
+          else
+            config.max_idle
+          end
+
+        {:noreply, {config, %{state | track: queue}}, timeout}
+    end
+  end
 
-    case HTTPoison.get(@track_endpoint, [], params: [data: data]) do
-      {:ok, %HTTPoison.Response{status_code: 200, body: "1"}} ->
-        :ok
+  def handle_cast({:engage, _event} = event, {config, state}) do
+    case Queue.push(state.engage, event) do
+      :dropped ->
+        new_state = Map.update!(state, :engage_dropped, &(&1 + 1))
+        {:noreply, {config, new_state}, 0}
+
+      {:ok, queue} ->
+        timeout =
+          if Queue.length(queue) >= config.batch_size do
+            0
+          else
+            config.max_idle
+          end
+
+        {:noreply, {config, %{state | engage: queue}}, timeout}
+    end
+  end
 
-      other ->
-        Logger.warn(
-          "Problem tracking Mixpanel event: #{inspect(event)}, #{inspect(properties)} Got: #{
-            inspect(other)
-          }"
-        )
+  def handle_info(:timeout, {%{batch_size: batch_size} = config, state}) do
+    new_state =
+      state
+      |> report_dropped()
+      |> engage_batch(batch_size, config.token)
+      |> track_batch(batch_size, config.token)
```

Comment:
When one of the queues is full both get sent to the mixpanel. Is it by design?

Also, I notice that `track_batch` for "engage" and "track" events are very similar, and that `handle_cast` for them is very similar too. Does it make sense to DRY them up, or maybe even have 2 GenServers: one per queue?

Reply:
I started writing a completely new Mixpanel client with a clean process architecture, and using Mint as the HTTP client. But it became too risky to introduce into production before Nov 16, so I tried patching up the existing client with (relatively) small changes. So yes, it's taking some shortcuts for simplicity, and this is one of them. In particular, I didn't want to change the process model, that's why I decided to stick with one GenServer.

I'll go over it one more time to see if I can remove some duplication.

Reply:
Got it 👍🏻

Reply:
> I'll go over it one more time to see if I can remove some duplication.

It doesn't seem worth it in these circumstances.